File: examples/lwjgl-map/src/main/java/org/maplibre/nativeffi/examples/lwjglmap/Shell.java
package org.maplibre.nativeffi.examples.lwjglmap;

import static org.lwjgl.glfw.GLFW.glfwPollEvents;
import static org.lwjgl.glfw.GLFW.glfwSetFramebufferSizeCallback;
import static org.lwjgl.glfw.GLFW.glfwSetWindowContentScaleCallback;
import static org.lwjgl.glfw.GLFW.glfwSetWindowSizeCallback;
import static org.lwjgl.glfw.GLFW.glfwWaitEventsTimeout;
import static org.lwjgl.glfw.GLFW.glfwWindowShouldClose;

import java.util.Set;
import org.maplibre.nativeffi.render.RenderBackend;

final class Shell {
  private static final int INITIAL_WIDTH = 960;
  private static final int INITIAL_HEIGHT = 640;
  private static final double IDLE_WAIT_SECONDS = 0.004;

  private Shell() {}

  static void run(RenderTargetMode mode, Set<RenderBackend> backends) throws Exception {
    try (var graphics =
        GraphicsContext.create("MapLibre LWJGL Map", INITIAL_WIDTH, INITIAL_HEIGHT, backends)) {
      var viewport = new ViewportHolder(Viewport.read(graphics.window()));
      viewport.value.log("initial viewport");
      try (var mapState = MapState.create(graphics, viewport.value, mode);
          var input =
              new InputController(graphics.window(), mapState.map(), mapState::requestRender)) {
        System.out.println("render target: " + mode.cliName());
        System.out.println("render target status: " + mode.status());
        InputController.printControls();
        installResizeCallbacks(graphics.window(), viewport);
        // TODO(map-example-spec): Replace poll-and-wait with a display-paced host loop. See Frame
        // loop.
        while (!glfwWindowShouldClose(graphics.window())) {
          glfwPollEvents();
          if (viewport.consumeChanged()) {
            viewport.value.log("resized viewport");
            if (!viewport.value.empty()) {
              graphics.resize(viewport.value);
              mapState.resize(viewport.value);
            }
          }
          if (viewport.value.empty()) {
            glfwWaitEventsTimeout(IDLE_WAIT_SECONDS);
            continue;
          }
          var rendered = mapState.step();
          if (!rendered) {
            glfwWaitEventsTimeout(IDLE_WAIT_SECONDS);
          }
        }
      }
    }
  }

  private static void installResizeCallbacks(long window, ViewportHolder viewport) {
    glfwSetWindowSizeCallback(window, (ignored, width, height) -> viewport.update(window));
    glfwSetFramebufferSizeCallback(window, (ignored, width, height) -> viewport.update(window));
    glfwSetWindowContentScaleCallback(window, (ignored, xScale, yScale) -> viewport.update(window));
  }

  private static final class ViewportHolder {
    private Viewport value;
    private boolean changed;

    ViewportHolder(Viewport value) {
      this.value = value;
    }

    void update(long window) {
      var next = Viewport.read(window);
      if (!next.equals(value)) {
        value = next;
        changed = true;
      }
    }

    boolean consumeChanged() {
      var result = changed;
      changed = false;
      return result;
    }
  }
}

File: examples/lwjgl-map/src/main/kotlin/org/maplibre/nativeffi/examples/lwjglmap/Shell.kt
package org.maplibre.nativeffi.examples.lwjglmap

import org.lwjgl.glfw.GLFW.glfwPollEvents
import org.lwjgl.glfw.GLFW.glfwPostEmptyEvent
import org.lwjgl.glfw.GLFW.glfwSetFramebufferSizeCallback
import org.lwjgl.glfw.GLFW.glfwSetWindowContentScaleCallback
import org.lwjgl.glfw.GLFW.glfwSetWindowSizeCallback
import org.lwjgl.glfw.GLFW.glfwWaitEventsTimeout
import org.lwjgl.glfw.GLFW.glfwWindowShouldClose
import org.maplibre.nativeffi.render.RenderBackend

/** The GLFW-thread shell that owns the window, graphics context, and render session. */
internal object Shell {
  private const val INITIAL_WIDTH = 960
  private const val INITIAL_HEIGHT = 640
  private const val IDLE_WAIT_SECONDS = 0.004

  fun run(mode: RenderTargetMode, backends: Set<RenderBackend>) {
    GraphicsContext.create("MapLibre LWJGL Map", INITIAL_WIDTH, INITIAL_HEIGHT, backends).use {
      graphics ->
      val initialViewport = Viewport.read(graphics.window())
      val viewport = ViewportHolder(initialViewport)
      initialViewport.log("initial viewport")
      val renderRequest = RenderRequest()

      MapState.create(initialViewport, ::glfwPostEmptyEvent).use { state ->
        renderLoop(graphics, mode, viewport, state, renderRequest)
      }
    }
  }

  private fun renderLoop(
    graphics: GraphicsContext,
    mode: RenderTargetMode,
    viewport: ViewportHolder,
    state: MapState,
    renderRequest: RenderRequest,
  ) {
    val target = RenderTarget.attach(graphics, state.map, viewport.value, mode)
    try {
      InputController(graphics.window(), state, renderRequest) { viewport.value }
        .use {
          println("render target: ${mode.cliName()}")
          println("render target status: ${mode.status()}")
          InputController.printControls()
          installResizeCallbacks(graphics.window(), viewport)
          while (!glfwWindowShouldClose(graphics.window())) {
            glfwPollEvents()
            state.drainNotifications(renderRequest)
            if (viewport.consumeChanged()) {
              viewport.value.log("resized viewport")
              if (!viewport.value.empty()) {
                graphics.resize(viewport.value)
                state.resize(viewport.value)
                target.resize(viewport.value)
                renderRequest.set()
              }
            }
            if (viewport.value.empty()) {
              glfwWaitEventsTimeout(IDLE_WAIT_SECONDS)
              continue
            }
            var rendered = false
            if (renderRequest.consume()) {
              rendered = render(target)
              if (!rendered) renderRequest.set()
            }
            if (!rendered) glfwWaitEventsTimeout(IDLE_WAIT_SECONDS)
          }
        }
    } finally {
      target.close()
    }
  }

  private fun render(target: RenderTarget): Boolean =
    if (target.needsMetalAutoreleasePool()) {
      MacObjectiveC.autoreleasePool().use { target.renderUpdate() }
    } else {
      target.renderUpdate()
    }

  private fun installResizeCallbacks(window: Long, viewport: ViewportHolder) {
    glfwSetWindowSizeCallback(window) { _, _, _ -> viewport.update(window) }
    glfwSetFramebufferSizeCallback(window) { _, _, _ -> viewport.update(window) }
    glfwSetWindowContentScaleCallback(window) { _, _, _ -> viewport.update(window) }
  }

  /** GLFW delivers every resize callback on the thread that polls. */
  private class ViewportHolder(var value: Viewport) {
    private var changed = false

    fun update(window: Long) {
      val next = Viewport.read(window)
      if (next != value) {
        value = next
        changed = true
      }
    }

    fun consumeChanged(): Boolean {
      val result = changed
      changed = false
      return result
    }
  }
}
